src/nvim/ex_getln.c

```diff
@@ -2258,11 +2280,30 @@ void putcmdline(int c, int shift)
 {
   if (cmd_silent)
     return;
-  msg_no_more = TRUE;
-  msg_putchar(c);
-  if (shift)
-    draw_cmdline(ccline.cmdpos, ccline.cmdlen - ccline.cmdpos);
-  msg_no_more = FALSE;
+  if (!cmdline_external) {
+    msg_no_more = true;
+    msg_putchar(c);
+    if (shift) {
+      draw_cmdline(ccline.cmdpos, ccline.cmdlen - ccline.cmdpos);
+    }
+    msg_no_more = false;
+  } else {
+      char_u *p;
```

Comment:
instead of modifying the text this could be a separate event `["cmdline_char", char, shift]`, This would allow the gui to change the display by showing it in a different color etc.